src/RadixDlt.NetworkGateway.PostgresIntegration/Services/ExtensionsQuerier.cs

```diff
@@ -0,0 +1,177 @@
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+using Dapper;
+using Microsoft.EntityFrameworkCore;
+using RadixDlt.NetworkGateway.Abstractions;
+using RadixDlt.NetworkGateway.Abstractions.Numerics;
+using RadixDlt.NetworkGateway.GatewayApi.Exceptions;
+using RadixDlt.NetworkGateway.GatewayApi.Services;
+using RadixDlt.NetworkGateway.PostgresIntegration.Models;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using GatewayModel = RadixDlt.NetworkGateway.GatewayApiSdk.Model;
+
+namespace RadixDlt.NetworkGateway.PostgresIntegration.Services;
+
+internal class ExtensionsQuerier : IExtensionsQuerier
+{
+    private readonly ReadOnlyDbContext _dbContext;
+    private readonly IDapperWrapper _dapperWrapper;
+
+    private record ResourceOwnersViewModel(long Id, EntityAddress EntityAddress, string Balance);
+
+    public ExtensionsQuerier(ReadOnlyDbContext dbContext, IDapperWrapper dapperWrapper)
+    {
+        _dbContext = dbContext;
+        _dapperWrapper = dapperWrapper;
+    }
+
+    public async Task<GatewayModel.ResourceOwnersResponse> ResourceOwners(
+        EntityAddress resourceAddress,
+        int limit,
+        GatewayModel.ResourceOwnersCursor? cursor,
+        CancellationToken token = default)
+    {
+        var resourceEntity = await _dbContext
+            .Entities
+            .AnnotateMetricName()
+            .FirstOrDefaultAsync(e => e.Address == resourceAddress, token);
+
+        if (resourceEntity == null)
+        {
+            throw new EntityNotFoundException(resourceAddress.ToString());
+        }
+
+        if (!resourceEntity.Address.IsResource)
+        {
+            throw new InvalidEntityException(resourceEntity.Address.ToString());
+        }
+
+        var totalCount = await _dbContext.ResourceOwners.CountAsync(x => x.ResourceEntityId == resourceEntity.Id, token);
+
+        var cd = new CommandDefinition(
+            commandText: @"
+SELECT
+    ro.id as Id,
+    e.address AS EntityAddress,
+    CAST(ro.balance AS text) AS Balance
+FROM resource_owners ro
+INNER JOIN entities e
+ON ro.entity_id = e.id
+WHERE ro.resource_entity_id = @resourceEntityId
+  AND (@balanceBoundary is null OR ((ro.balance, ro.id) < (Cast(@balanceBoundary AS numeric(1000,0)), @idBoundary)))
+ORDER BY (ro.balance, ro.entity_id) DESC
+LIMIT @limit",
+            parameters: new
+            {
+                resourceEntityId = resourceEntity.Id,
+                balanceBoundary = cursor?.BalanceBoundary,
```

Comment:
Guess it would be easier to default balance boundary to `0`? No extra if-o-logy in the query itself.

Reply:
Ah, we'd have to default it to "max". I think `TokenAmount` should actually define such a construct.

Reply:
It seems that scrypto defines `Decimal.MAX` as `I192::MAX`. This could be encoded in our `TokenAmount`. Should have no impact on performance and cleaner code is of higher priority if you ask me.